### PR TITLE
fix(pipeline): support for API changes

### PR DESCRIPTION
**Pull request by jbrockopp**
Dependent on https://github.com/go-vela/server/pull/615

Part of the effort for https://github.com/go-vela/community/issues/460

> Shout out to Vader for all the help with this change! ❤️ 

This change updates the UI to support the updated API endpoints for `pipelines`.

Previously, when fetching a pipeline from the API, we'd receive the raw pipeline in a `string` format:

i.e.

```yaml
version: "1"

steps:
  - name: sleep
    image: alpine:latest
    commands:
      - sleep 10
```

Now, the API returns a pipeline record with multiple fields including the base64 encoded raw pipeline (`data`):

```json
{
  "id": 1,
  "repo_id": 1,
  "commit": "379012cab91195b19b6e8f98932d7d74e10ab002",
  "flavor": "",
  "platform": "",
  "ref": "refs/heads/sleep",
  "type": "yaml",
  "version": "1",
  "external_secrets": false,
  "internal_secrets": false,
  "services": false,
  "stages": false,
  "steps": true,
  "templates": false,
  "data": "dmVyc2lvbjogIjEiCgpzdGVwczoKICAtIG5hbWU6IHNsZWVwCiAgICBpbWFnZTogYWxwaW5lOmxhdGVzdAogICAgY29tbWFuZHM6CiAgICAgIC0gc2xlZXAgMTA="
}
```

Additionally, we used to fetch pipelines via a `ref` query parameter sent to the API

http://localhost:8888/JordanBrockopp/vela-play/1/pipeline?ref=379012cab91195b19b6e8f98932d7d74e10ab002

Now, the `ref` query parameter has been moved into the URL path:

http://localhost:8888/JordanBrockopp/vela-play/1/pipeline/379012cab91195b19b6e8f98932d7d74e10ab002

Here's a screenshot of what I see locally:

image

Co-authored-by: @plyr4

### CODE REVIEW

**Comment by plyr4** [comment]
two fixes: (1/2) i believe you want the build navigation tab for "Pipeline" to point to the page that does not require a BuildNumber

https://github.com/go-vela/ui/pull/539/files#diff-8eaa47761f7ea99f8c950be294b2292cbd55d36dd800e9f50f11753e2c7aa686R376

```diff
-            , Tab "Pipeline" currentPage (Pages.BuildPipeline org repo buildNumber ref pipeline.expand pipeline.focusFragment) False
+            , Tab "Pipeline" currentPage (Pages.Pipeline org repo ref pipeline.expand pipeline.focusFragment) False 
```

then see [this comment]()

File: src/elm/Routes.elm [hunk 1]
```diff
@@ -179,10 +179,10 @@ routeToUrl route =
             "/" ++ org ++ "/" ++ repo ++ "/" ++ buildNumber ++ "/services" ++ Maybe.withDefault "" lineFocus
 
         BuildPipeline org repo buildNumber ref expand lineFocus ->
-            "/" ++ org ++ "/" ++ repo ++ "/" ++ buildNumber ++ "/pipeline" ++ (UB.toQuery <| List.filterMap identity <| [ maybeToQueryParam ref "ref", maybeToQueryParam expand "expand" ]) ++ Maybe.withDefault "" lineFocus
+            "/" ++ org ++ "/" ++ repo ++ "/" ++ buildNumber ++ "/pipeline" ++ (UB.toQuery <| List.filterMap identity <| [ maybeToQueryParam expand "expand" ]) ++ Maybe.withDefault "" lineFocus
 
         Pipeline org repo ref expand lineFocus ->
-            "/" ++ org ++ "/" ++ repo ++ "/pipeline" ++ (UB.toQuery <| List.filterMap identity <| [ maybeToQueryParam ref "ref", maybeToQueryParam expand "expand" ]) ++ Maybe.withDefault "" lineFocus
+            "/" ++ org ++ "/" ++ repo ++ "/pipeline" ++ (UB.toQuery <| List.filterMap identity <| [ maybeToQueryParam expand "expand" ]) ++ Maybe.withDefault "" lineFocus
```
Comment (plyr4) [inline src/elm/Routes.elm:185]: (2/2) see first comment

this is the second half of the fix

```suggestion
            "/" ++ org ++ "/" ++ repo ++ "/" ++ ref ++ "/pipeline" ++ (UB.toQuery <| List.filterMap identity <| [ maybeToQueryParam expand "expand" ]) ++ Maybe.withDefault "" lineFocus
```

**Comment by plyr4** [comment]
> two fixes: (1/2) i believe you want the build navigation tab for "Pipeline" to point to the page that does not require a BuildNumber
> 
> https://github.com/go-vela/ui/pull/539/files#diff-8eaa47761f7ea99f8c950be294b2292cbd55d36dd800e9f50f11753e2c7aa686R376
> 
> ```diff
> -            , Tab "Pipeline" currentPage (Pages.BuildPipeline org repo buildNumber ref pipeline.expand pipeline.focusFragment) False
> +            , Tab "Pipeline" currentPage (Pages.Pipeline org repo ref pipeline.expand pipeline.focusFragment) False 
> ```
> 
> then see [this comment](https://github.com/go-vela/ui/pull/539/files#r857925808)

on second thought, if you keep the nav tab pointing to the same location you can simply update `routeToUrl` 

```diff
-        BuildPipeline org repo buildNumber ref expand lineFocus ->
-            "/" ++ org ++ "/" ++ repo ++ "/" ++ buildNumber ++ "/pipeline" ++ (UB.toQuery <| List.filterMap identity <| [ maybeToQueryParam expand "expand" ]) ++ Maybe.withDefault "" lineFocus

+        BuildPipeline org repo _ ref expand lineFocus ->
+            "/" ++ org ++ "/" ++ repo ++ "/" ++ ref ++ "/pipeline" ++ (UB.toQuery <| List.filterMap identity <| [ maybeToQueryParam expand "expand" ]) ++ Maybe.withDefault "" lineFocus

-        Pipeline org repo ref expand lineFocus ->
-            "/" ++ org ++ "/" ++ repo ++ "/pipeline" ++ (UB.toQuery <| List.filterMap identity <| [ maybeToQueryParam expand "expand" ]) ++ Maybe.withDefault "" lineFocus

+        Pipeline org repo ref expand lineFocus ->
+            "/" ++ org ++ "/" ++ repo ++ "/" ++ ref ++ "/pipeline" ++ (UB.toQuery <| List.filterMap identity <| [ maybeToQueryParam expand "expand" ]) ++ Maybe.withDefault "" lineFocus

```

that should work better

**Review by cognifloyd — APPROVED** [review]
I don't know much about elm and cypress, but from reading it looks like a logical change - at least the endpoints look right.

I approve with the caveat that someone else who knows what they're doing really should check this out...

**Pull request merged**